### ISSUE TITLE
Incomplete transfer of module to device (only with 0.101.3)

### ISSUE THREAD

**Issue by dje-dev** [dotnet/TorchSharp#1185]
Upon upgrading to 0.101.3 a previously working network now throws exception upon calling forward. The module is constructed on CPU with multiple layers in constructor and as normal, RegisterComponents is called. Module is subsequently moved to cuda. Most layers successfully are transferred, but at least one (LayerNorm) remains on cpu. This causes exception below.

Downgrading back to 0.101.2 immediately fixed the problem.
  
Unhandled exception. System.Runtime.InteropServices.ExternalException (0x80004005): Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument weight in method wrapper_CUDA__native_layer_norm)

**Comment by NiklasGustafsson** [comment]
@shaltielshmid -- I think this one will require your attention.

**Comment by shaltielshmid** [comment]
Can you please send sample code to replicate this behavior?

**Comment by shaltielshmid** [comment]
Okay, after some digging, I found the root of the problem to be from here: https://github.com/dotnet/TorchSharp/pull/1152

In TorchSharp 0.101.3 the LayerNorm class was changed to act like torch.nn.functional.
In the constructor, the LayerNorm module calls the base constructor with a "IntPtr.Zero" handle. 
When `module.to()` gets called, then it goes through all the registered modules and calls `.to()` on them as well. Because LayerNorm's handle is zero, so `LayerNorm.handle.IsInvalid` always returns true and therefore it is not registered as a module. 

You can see the problematic line [here](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/NN/Normalization/LayerNorm.cs#L23).

@NiklasGustafsson Is there any reason to register the LayerNorm module with an invalid handle? Ideally we should just pass a string with the name of the module (either default it to "LayerNorm" or require it in the constructor).

**Comment by GilesBathgate** [comment]
@shaltielshmid if `base(IntPtr.Zero, IntPtr.Zero)` is problematic then it can easily be replaced with a named module base constructor. I chose to use IntPtr.Zero as this seems to be the way its done elsewhere, when there is no module handle required for calls to native functions. How does module.to work with user defined modules?

**Comment by GilesBathgate** [comment]
@NiklasGustafsson Perhaps I am meant to call `base(IntPtr.Zero, IntPtr.Zero ownsHandle: false)` ?

**Comment by shaltielshmid** [comment]
@GilesBathgate I'm not familiar with other modules that call `base(IntPtr.Zero, IntPtr.Zero)`, can you send an example?

The problem with IntPtr.Zero is that the module is considered invalid and then it is not registered as a submodule of the main module. Aside from the `module.to()` issue this also leads to other problems, such as the `LayerNorm.weight` and `LayerNorm.bias` parameters not being returned in the `module.parameters()` function of the parent module. This means that the weights aren't passed to the optimizer and they aren't being tuned.

**Comment by GilesBathgate** [comment]
@shaltielshmid Literally every named user module, unless I am misunderstanding: https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/NN/Module.cs#L965

**Comment by shaltielshmid** [comment]
@GilesBathgate In that case it's correct, since the constructor assigns the handle and boxedModule separately.

**Comment by NiklasGustafsson** [comment]
@shaltielshmid -- is the check for a valid handle on Module new? Every module that isn't implemented in native code will have an invalid native module handle.

**Comment by shaltielshmid** [comment]
@NiklasGustafsson No, it's always been there (well, at least since I started using TorchSharp) - located [here](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/NN/Module.cs#L707). 

But every module is given a valid handle in the constructor of a custom module [here](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/NN/Module.cs#L981).

**Comment by GilesBathgate** [comment]
@shaltielshmid I see, yes, if I call `base(IntPtr.Zero, IntPtr.Zero)` I should also do `handle = new HType(res, true);` That was not what I expected. Fix would be like you say to call the named constructor., or add a new constructor that allows no name but inits handle accordingly.

**Comment by shaltielshmid** [comment]
@GilesBathgate Agreed.
I would vote for the named constructor option, since the Module is a custom module and probably best not to duplicate the code which creates the handle.

**Comment by GilesBathgate** [comment]
I do wonder if Unnamed modules are a thing, and whether giving LayerNorm a name is compatible with PyTorch.

**Comment by NiklasGustafsson** [comment]
The names really don't have any functional significance.

**Comment by shaltielshmid** [comment]
To quote the documentation:
```
<param name="name">The name of the module. Useful for debugging purposes, mostly.</param>
```

**Comment by GilesBathgate** [comment]
@NiklasGustafsson Oh, TorchSharp names are not pytorch module names which live in the container?

```python
>>> import torch
>>> l=torch.nn.LayerNorm((1,1))
>>> list(l.named_modules())
[('', LayerNorm((1, 1), eps=1e-05, elementwise_affine=True))]
>>>
>>>
>>> s=torch.nn.Sequential()
>>> s.add_module("Its Name",l)
>>> list(s.named_modules())
[('', Sequential(
  (Its Name): LayerNorm((1, 1), eps=1e-05, elementwise_affine=True)
)), ('Its Name', LayerNorm((1, 1), eps=1e-05, elementwise_affine=True))]

```

**Comment by NiklasGustafsson** [comment]
No, the module "name" is not the same as the name by which a submodule is registered.

**Comment by NiklasGustafsson** [comment]
In the not-yet-merged branch called `unit`, the reimplemented modules do this:

```C#
internal Linear(long inputSize, long outputSize, bool hasBias = true, Device? device = null, ScalarType? dtype = null) : base(nameof(Linear))
```

**Comment by NiklasGustafsson** [comment]
Okay, I'll fix it and do a fast-follow release.

**Comment by NiklasGustafsson** [comment]
I also found that it's not sufficient -- you also need to override the `_to()` overloads and set the properties like this:

```C#
            protected internal override Module _to(ScalarType dtype)
            {
                var result = base._to(dtype) as LayerNorm;
                result!._weight = result.get_parameter("weight");
                if (result.has_parameter("bias")) {
                    result!._bias = result.get_parameter("bias");
                }
                return this;
            }
```

Because the new base class logic only looks at fields, so we have to set the property in the derived class.

**Comment by shaltielshmid** [comment]
@NiklasGustafsson 
Why not just use the ComponentName attribute? That should solve the issue

**Comment by NiklasGustafsson** [comment]
> @NiklasGustafsson ok, so that sounds like a good pattern. Any way to can catch misuse such as what I implemented (Unit tests I imagine)

I suspect that validating that the passed `handle` argument is valid is necessary. I'm a little worried about what that may break, though.

**Comment by NiklasGustafsson** [comment]
> @NiklasGustafsson Why not just use the ComponentName attribute? That should solve the issue

No, you're supposed to go through the property setter.

**Comment by GilesBathgate** [comment]
@NiklasGustafsson Shouldn't the IntPtr,IntPtr base method just be private?

**Comment by shaltielshmid** [comment]
> > @NiklasGustafsson Why not just use the ComponentName attribute? That should solve the issue
> 
> No, you're supposed to go through the property setter.

Why? I don't see a specific reason to.

**Comment by NiklasGustafsson** [comment]
From a pattern perspective -- make sure to catch any specialized logic contained in the property setters. It may not be necessary on all modules.

**Comment by shaltielshmid** [comment]
> From a pattern perspective -- make sure to catch any specialized logic contained in the property setters. It may not be necessary on all modules.

In the _to() override you wrote you used the field setter.

**Comment by shaltielshmid** [comment]
Also, as a side note: why use property setters with conditional register instead of just using the default RegisterComponents()?

**Comment by GilesBathgate** [comment]
@shaltielshmid FYI: https://github.com/dotnet/TorchSharp/pull/1152#discussion_r1406436320

**Comment by shaltielshmid** [comment]
> @shaltielshmid FYI: [#1152 (comment)](https://github.com/dotnet/TorchSharp/pull/1152#discussion_r1406436320)

Thanks!

**Comment by shaltielshmid** [comment]
> > > From a pattern perspective -- make sure to catch any specialized logic contained in the property setters. It may not be necessary on all modules.
> > 
> > 
> > In the _to() override you wrote you used the field setter.
> 
> Which I have since fixed... :-)

Great!
One more suggestion:

Since `Module.to()` modifies in place, perhaps instead of storing the result module, just call base. For example:

```cs
            protected internal override Module _to(ScalarType dtype)
            {
                base._to(dtype);
                this.weight = base.get_parameter("weight");
                if (base.has_parameter("bias")) {
                    this.bias = base.get_parameter("bias");
                }
                return this;
            }
```

**Comment by NiklasGustafsson** [comment]
Makes sense.

**Comment by NiklasGustafsson** [comment]
I take back everything I said about overriding `_to()` -- the property setter pattern was defined before ComponentName came into place, but with it, the base class takes care of it.

**Comment by NiklasGustafsson** [comment]
But here's a question for you, @shaltielshmid... In Module.cs, lines 253-255, the following code exists:

```C#
                        using (var d = torch.no_grad())
                            p = new Parameter(param.to(paramType, device ?? param.device, disposeAfter: true), requiresGrad);
                        ConditionallyRegisterParameter(name, p);
```

What happens if there's an active dispose scope? Shouldn't the parameter be detached from the dispose scope when created?

**Comment by shaltielshmid** [comment]
Good catch! 
I think that was an issue with the previous implementation as well.
